File: day5.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)
#define ll long long
#define ii pair<ll, ll>

void part1() {
	ll ans = 0;
	vector<ll> seeds;
	INPUT(line) {
		stringstream ss(line);
		string tmp; ss >> tmp;
		
		ll seed;
		while (ss >> seed)
			seeds.push_back(seed);
	}

	for (int i = 0; i < 7; ++i) {
		string tmp; getline(cin, tmp);
		vector<ll> nxt;
		INPUT(line) {
			stringstream ss(line);
			ll d, s, rng;
			ss >> d >> s >> rng;

			for (ll& seed : seeds) {
				if (seed >= s && seed < s + rng) {
					nxt.push_back(d + (seed - s));
					seed = -1;
				}
			}
		}
		for (ll seed : seeds)
			if (seed != -1)
				nxt.push_back(seed);

		seeds = vector<ll>(begin(nxt), end(nxt));
		nxt.clear();
	}
	cout << *min_element(begin(seeds), end(seeds)) << '\n';
}

void part2() {
	vector<ii> seeds;
	INPUT(line) {
		stringstream ss(line);
		string tmp; ss >> tmp;

		ll start;
		while (ss >> start) {
			ll rng; ss >> rng;
			seeds.push_back({ start, start+rng-1 });
		}
	}

	for (int i = 0; i < 7; ++i) {
		string tmp; getline(cin, tmp);
		vector<ii> nxt;
		INPUT(line) {
			stringstream ss(line);
			ll d, s, rng;
			ss >> d >> s >> rng;

			// extra ranges when one splits into two
			vector<ii> q;
			for (auto& [l, r] :seeds) {
				if (l == -1) continue;
				ll sEnd = s + rng - 1;
				// not in range
				if (r < s || l > sEnd) continue;
				if (l <= s && r >= sEnd) {
					// fully encompasses range
					nxt.push_back({ d, d + rng });
					if (sEnd < r)
						q.push_back({ sEnd + 1, r });
					if (l < s)
						r = s - 1;
					else
						l = -1;
					break;
				}

				// partial overlap
				ll L = max(l, s), R = min(r, sEnd);
				nxt.push_back({ d+L-s, d+R-s });
				if (l < L)
					r = L - 1;
				else if (r > R)
					l = R + 1;
				else
					l = -1;
			}
			for (ii rng : q)
				seeds.push_back(rng);
		}
		// build resulting set of ranges
		for (auto& [l, r] : seeds)
			if (l != -1)
				nxt.push_back({ l, r });

		seeds = vector<ii>(begin(nxt), end(nxt));
		nxt.clear();
	}
	cout << min_element(begin(seeds), end(seeds))->first << '\n';
}

void main() {
	part1();
	part2();
}

File: day5.py
def part1():
    seeds = [int(s) for s in input().split()[1:]]
    input()
    for i in range(7):
        input()
        new_seeds = []
        while line := input():
            dest, source, size = map(int, line.split())
            for i in range(len(seeds)):
                if not seeds[i]:
                    continue
                if seeds[i] >= source and seeds[i] < source+size:
                    new_seeds.append(dest+seeds[i]-source)
                    seeds[i] = None
        for seed in seeds:
            if seed:
                new_seeds.append(seed)
        seeds = new_seeds
    print(min(*seeds))

def part2():
    seeds = [int(s) for s in input().split()[1:]]
    ranges = [[seeds[i], seeds[i]+seeds[i+1]-1] for i in range(0, len(seeds), 2)]
    input()
    for i in range(7):
        input()
        new_ranges = []
        while line := input():
            dest, source, size = map(int, line.split())
            for i in range(len(ranges)):
                if not ranges[i]:
                    continue
                
                l, r = ranges[i]
                if r < source or l >= source+size:
                    continue
                if l >= source and r < source+size:
                    new_ranges.append([dest+l-source, dest+r-source])
                    ranges[i] = None
                    continue
                L, R = max(l, source), min(r, source+size-1)
                new_ranges.append([dest+L-source, dest+R-source])
                if l < L:
                    ranges[i][1] = L-1
                else:
                    ranges[i] = None
                if r > R:
                    ranges.append([R+1, r])

        for rng in ranges:
            if rng:
                new_ranges.append(rng)
        ranges = new_ranges
    print(min(*[l for l,r in ranges]))

part1()
part2()